### PR TITLE
Fire Hydrant Additions.

### PR DESCRIPTION
**Pull request by joncatanio**
Simple fire hydrant additions with models! Woohoo! We are doing kind of hacky physics but if we like the feel of it that's great. Definitely pull this down and check it out. Also take a look at my vector math and see if that looks alright.

### CODE REVIEW

File: levels/grandmas-awakening.json [hunk 1]
```diff
@@ -923,6 +927,125 @@
...
+            "name": "FireHydrantPhysicsComponent"
+         },
+         "render-component": {
+            "name": "WallRenderComponent",
+            "shape": "FireHydrant",
+            "shader": "Red",
+            "material": "Obsidian"
+         },
+         "deliverable": false
+      }
```
Comment (rgarmsen2295) [inline levels/grandmas-awakening.json:1085]: Can you change the Fire Hydrants to have a Red material and use the phong shader? I'm going to remove the dummy red/blue/green shaders with the lighting stuff and we need to start using materials again (sadly :P )

File: src/BoundingBox.cpp [hunk 2]
```diff
@@ -36,6 +36,27 @@ bool BoundingBox::checkIntersection(BoundingBox& other) {
 		(this->min_.z <= other.max_.z && this->max_.z >= other.min_.z);
 }
 
+glm::vec3 BoundingBox::calcReflNormal(BoundingBox& other) {
+   glm::vec3 normal;
+   float epsilon = 1.5;
```
Comment (rgarmsen2295) [inline src/BoundingBox.cpp:41]: Didn't realize the epsilon was that big, damn.  Makes sense though.

File: src/FireHydrantPhysicsComponent.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+   }
+
+   if (objHit != nullptr) {
+      GameObjectType objTypeHit = objHit->type;
+      
+      if (objTypeHit == GameObjectType::PLAYER) {
+         glm::vec3 reactDir = objHit->direction;
+         glm::vec3 rotAxis = glm::cross(objHit->direction, glm::vec3(0, 1, 0));
+         // Give the direction of the hydrant a vertical component.
+         reactDir = glm::rotate(reactDir, (float)(M_PI / 6.0f), rotAxis);
```
Comment (rgarmsen2295) [inline src/FireHydrantPhysicsComponent.cpp:42]: Houston, we have liftoff!!!

**Review by rgarmsen2295 — APPROVED** [review]
Looks good, nothing crazy from code to math, good stuff.  Didn't pull down, but I trust it runs :)

Only req. is a minor change with materials vs. shaders

File: levels/grandmas-awakening.json (hunk 1 above)
Comment (joncatanio) [reply levels/grandmas-awakening.json:1085]: Do we have a red material? I can always look one up. I think I have one actually from my old 471 stuff. Will look into it.

File: src/BoundingBox.cpp (hunk 2 above)
Comment (joncatanio) [reply src/BoundingBox.cpp:41]: Yeaahhh, there is no visual difference but this seemed to be the least buggy

File: src/FireHydrantPhysicsComponent.cpp (hunk 3 above)
Comment (joncatanio) [reply src/FireHydrantPhysicsComponent.cpp:42]: Lol yup! Love this shit!

File: levels/grandmas-awakening.json (hunk 1 above)
Comment (rgarmsen2295) [reply levels/grandmas-awakening.json:1085]: Yeah, we'll need a red material, you can either leave as obsidian and I can fill in or just add one.

Comment (nurgan) [reply levels/grandmas-awakening.json:1085]: There is a webpage that lists a shitton of materials, I'll look it up if we need it

File: src/BoundingBox.cpp (hunk 2 above)
Comment (nurgan) [reply src/BoundingBox.cpp:41]: Yeah, the epsilon needs to really big, I was surprised by that too when I did the logic originally. We might want to ask Zoe at some point whats the best for that. We should also ask her if there is a not too complicated way to make the bouncing (for basic collisions it's fine) not axis aligned

File: src/FireHydrantPhysicsComponent.cpp (hunk 3 above)
Comment (nurgan) [reply src/FireHydrantPhysicsComponent.cpp:42]: We should definitely introduce a random component to the velocity. After we update the matrix math a rotation while it's in the air. (Same for the cookie btw)

**Review by nurgan — APPROVED** [review]
Made some comments, nothing that needs to be done now. Looks good. Pulled it, buildes and workes on my machine.

**Pull request merged**